larky-api/src/main/java/com/verygood/security/larkyapi/VersionedLarkyEngineImpl.java

```diff
@@ -1,227 +1,246 @@
 import java.lang.reflect.Method;
-import java.net.MalformedURLException;
-import java.net.URISyntaxException;
 import java.net.URL;
 import java.net.URLClassLoader;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
 import java.util.HashMap;
 import java.util.Map;
 import java.util.Set;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
+import java.util.stream.Stream;
 
+@Slf4j
 public class VersionedLarkyEngineImpl implements VersionedLarkyEngine {
 
-  final private Class engineClass;
-  final private Class scriptClass;
-  final private Class parseClass;
-  final private ScriptEngine engineInstanceObj;
-  final private String version;
-
-
-  // Instantiating the static map
-  private static Map<String, URL> larkyJarByVersion;
-  static {
-    detectVersions();
-  }
-
-  private static void detectVersions() {
-    larkyJarByVersion = new HashMap<>();
-    try {
-      PathMatchingResourcePatternResolver resolver =
-              new PathMatchingResourcePatternResolver(VersionedLarkyEngineImpl.class.getClassLoader());
-      // Get jars with format `larky-\d{>=1}.\d{>=1}.\d{>=1}-fat.jar`
-      Resource[] resources = resolver.getResources( // uses AntPathMatcher
-              "classpath*:larky-{\\d+}.{\\d+}.{\\d+}-fat.jar"
-      );
-      for (Resource resource: resources) {
-        String fileName = resource.getFilename();
-        URL fileURL = resource.getURL();
-
-        Pattern pattern = Pattern.compile("\\d+.\\d+.\\d+");
-        Matcher matcher = pattern.matcher(fileName);
-        if (matcher.find()) {
-          larkyJarByVersion.put(matcher.group(), fileURL);
+    final private Class engineClass;
+    final private Class scriptClass;
+    final private Class parseClass;
+    final private ScriptEngine engineInstanceObj;
+    final private String version;
+
+    private static IOException setupException = null;
+
+
+    // Instantiating the static map
+    private static Map<String, URL> larkyJarByVersion;
+
+    static {
+        detectVersions();
+    }
+
+    private static void detectVersions() {
+        larkyJarByVersion = new HashMap<>();
+        String larky_lib = System.getProperty("user.home") + "/.larky/lib"; // default dir
+
+        String larky_alt_lib = System.getenv("LARKY_LIB_HOME");
+        if (larky_alt_lib != null) {
+            larky_lib = larky_alt_lib;
+        }
+
+        try {
+            Stream<Path> paths = Files.walk(Paths.get(larky_lib));
+            paths
+                    // Get jars with format `larky-\d{>=1}.\d{>=1}.\d{>=1}-fat.jar`
+                    .filter(filePath -> filePath.toString().matches("^.*larky-\\d+.\\d+.\\d+-fat.jar$"))
+                    .forEach(filePath -> {
+                        try {
+                            String fileName = filePath.toString();
+                            URL fileURL = filePath.toUri().toURL();
+
+                            Pattern pattern = Pattern.compile("\\d+.\\d+.\\d+");
```

Comment:
Do not compile pattern on every loop cycle of foreach
https://www.baeldung.com/java-regex-pre-compile